grovedb/src/replication.rs

```diff
@@ -84,57 +105,78 @@ impl GroveDb {
+        let root_app_hash = self.root_hash(None, grove_version).value?;
+        if packed_global_chunk_id.len() == root_app_hash.len() {
+            global_chunk_ids.push(packed_global_chunk_id.to_vec());
+        } else {
+            global_chunk_ids.extend(unpack_nested_bytes(packed_global_chunk_id)?);
         }
 
-        let mut chunk_producer = ChunkProducer::new(&merk).map_err(|e| {
-            Error::CorruptedData(format!(
-                "failed to create chunk producer by prefix tx:{} with:{}",
-                hex::encode(chunk_prefix),
-                e
-            ))
-        })?;
-        let (chunk, _) = chunk_producer
-            .chunk(&chunk_id, grove_version)
-            .map_err(|e| {
-                Error::CorruptedData(format!(
-                    "failed to apply chunk:{} with:{}",
-                    hex::encode(chunk_prefix),
-                    e
-                ))
-            })?;
-        let op_bytes = utils::encode_vec_ops(chunk).map_err(|e| {
-            Error::CorruptedData(format!(
-                "failed to encode chunk ops:{} with:{}",
-                hex::encode(chunk_prefix),
-                e
-            ))
-        })?;
-        Ok(op_bytes)
+        let mut global_chunk_bytes: Vec<Vec<u8>> = vec![];
+        for global_chunk_id in global_chunk_ids {
+            let (chunk_prefix, root_key, tree_type, nested_chunk_ids) =
+                utils::decode_global_chunk_id(global_chunk_id.as_slice(), &root_app_hash)?;
+
+            let mut local_chunk_bytes: Vec<Vec<u8>> = vec![];
+
+            let merk = self
+                .open_transactional_merk_by_prefix(
+                    chunk_prefix,
+                    root_key,
+                    tree_type,
+                    tx.as_ref(),
+                    None,
+                    grove_version,
+                )
+                .value
+                .map_err(|e| {
+                    Error::CorruptedData(format!(
+                        "failed to open merk by prefix tx:{} with:{}",
+                        hex::encode(chunk_prefix),
+                        e
+                    ))
+                })?;
+            if merk.is_empty_tree().unwrap() {
+                local_chunk_bytes.push(vec![]);
+            } else {
+                let mut chunk_producer = ChunkProducer::new(&merk).map_err(|e| {
+                    Error::CorruptedData(format!(
+                        "failed to create chunk producer by prefix tx:{} with:{}",
+                        hex::encode(chunk_prefix),
+                        e
+                    ))
+                })?;
+                // Ensure we iterate once with vec![] if it's empty
+                let iter: Box<dyn Iterator<Item = Vec<u8>>> = if nested_chunk_ids.is_empty() {
+                    Box::new(std::iter::once(vec![]))
+                } else {
+                    Box::new(nested_chunk_ids.into_iter())
+                };
+                for chunk_id in iter {
```

Comment:
I'm not sure actually what will be better but `nested_chunk_ids.is_empty().then(|| vec![]).into_iter().chain(nested_chunk_ids.into_iter())` in place of `iter` could work instead of dynamic dispatch above